# erro

Uma Ferramenta [Sumor Cloud](https://sumor.cloud).  
[Documentação Adicional](https://sumor.cloud)
Uma biblioteca de tratamento de erro que suporta múltiplos idiomas, códigos de erro predefinidos e passagem de dados.

[![CI](https://github.com/sumor-cloud/error/actions/workflows/ci.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ci.yml)
[![Teste](https://github.com/sumor-cloud/error/actions/workflows/ut.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ut.yml)
[![Cobertura](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml)
[![Auditoria](https://github.com/sumor-cloud/error/actions/workflows/audit.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/audit.yml)

## Instalação

```bash
npm i @sumor/error --save
```

## Pré-requisitos

### Versão do Node.JS

Requer Node.JS versão 16.x ou superior

### Requerir módulo ES do Node.JS

Como este pacote é escrito em módulo ES,
por favor altere o seguinte código em seu arquivo `package.json`:

```json
{
  "type": "module"
}
```

## Uso

### Definir Erro

```js
import defineError from '@sumor/error'

const MeuErro = defineError({
  code: {
    USUÁRIO_NÃO_ENCONTRADO: 'Usuário não encontrado',
    USUÁRIO_EXISTENTE: 'Usuário {name} já existe'
  }
})

throw new MeuErro('USUÁRIO_NÃO_ENCONTRADO')
// saída: Error: Usuário não encontrado

throw new MeuErro('USUÁRIO_EXISTENTE', { name: 'Alice' })
// saída: Error: Usuário Alice já existe
```

### Multi-idioma

```js
import defineError from '@sumor/error'

const MeuErro = defineError({
  language: 'en', // idioma padrão
  code: {
    USUÁRIO_NÃO_ENCONTRADO: 'Usuário não encontrado',
    USUÁRIO_EXISTENTE: 'Usuário {name} já existe'
  },
  i18n: {
    zh: {
      USUÁRIO_NÃO_ENCONTRADO: 'Usuário não encontrado',
      USUÁRIO_EXISTENTE: 'Usuário {name} já existe'
    }
  }
})

const erro = new MeuErro('USUÁRIO_EXISTENTE', { name: 'Alice' })
erro.language = 'en' // alterar idioma do erro
console.log(erro)
// saída: Error: Usuário Alice já existe

erro.language = 'zh' // alterar idioma do erro
console.log(erro)
// saída: Error: Usuário Alice já existe
```

### Converter Erro para JSON

```js
import defineError from '@sumor/error'

const MeuErro = defineError({
  code: {
    USUÁRIO_NÃO_ENCONTRADO: 'Usuário não encontrado',
    USUÁRIO_EXISTENTE: 'Usuário {name} já existe'
  }
})

const erro = new MeuErro('USUÁRIO_EXISTENTE', { name: 'Alice' })
console.log(erro.json())
// saída: {"code":"USUÁRIO_EXISTENTE","message":"Usuário Alice já existe"}
```

### Erro Subjacente

```js
import defineError from '@sumor/error'

const MeuErro = defineError({
  code: {
    FALHA_VERIFICAÇÃO_CAMPO: 'Falha na verificação do campo',
    CAMPO_NÃO_PODE_FICAR_VAZIO: 'Campo {name} não pode ficar vazio',
    CAMPO_MUITO_LONGO: 'Campo {name} está muito longo'
  },
  i18n: {
    zh: {
      FALHA_VERIFICAÇÃO_CAMPO: 'Falha na verificação do campo',
      CAMPO_NÃO_PODE_FICAR_VAZIO: 'Campo {name} não pode ficar vazio',
      CAMPO_MUITO_LONGO: 'Campo {name} está muito longo'
    }
  }
})

const erro = new MeuErro('FALHA_VERIFICAÇÃO_CAMPO', {}, [
  new MeuErro('CAMPO_NÃO_PODE_FICAR_VAZIO', { name: 'nome de usuário' }),
  new MeuErro('CAMPO_MUITO_LONGO', { name: 'senha' })
])

console.log(erro.json())
/* 
saída: 
{
  "code":"FALHA_VERIFICAÇÃO_CAMPO",
  "message":"Falha na verificação do campo",
  "errors":[
    {
      "code":"CAMPO_NÃO_PODE_FICAR_VAZIO",
      "message":"Campo nome de usuário não pode ficar vazio"
    },{
      "code":"CAMPO_MUITO_LONGO",
      "message":"Campo senha está muito longo"
    }
  ]
}
*/

erro.language = 'zh'
console.log(erro.json())
/*
saída:
{
  "code":"FALHA_VERIFICAÇÃO_CAMPO",
  "message":"Falha na verificação do campo",
  "errors":[
    {
      "code":"CAMPO_NÃO_PODE_FICAR_VAZIO",
      "message":"Campo nome de usuário não pode ficar vazio"
    },{
      "code":"CAMPO_MUITO_LONGO",
      "message":"Campo senha está muito longo"
    }
  ]
}
*/
```

### Combinar Erro Padrão

```js
import defineError from '@sumor/error'

const MeuErro = defineError({
  code: {
    FALHA_VERIFICAÇÃO_CAMPO: 'Falha na verificação do campo',
    CAMPO_NÃO_PODE_FICAR_VAZIO: 'Campo {name} não pode ficar vazio',
    CAMPO_MUITO_LONGO: 'Campo {name} está muito longo'
  }
})

const erro = new MeuErro('FALHA_VERIFICAÇÃO_CAMPO', {}, [
  new MeuErro('CAMPO_NÃO_PODE_FICAR_VAZIO', { name: 'nome de usuário' }),
  new MeuErro('CAMPO_MUITO_LONGO', { name: 'senha' }),
  new Error('Erro Desconhecido')
])

console.log(erro.json())
/*
saída:
{
  "code":"FALHA_VERIFICAÇÃO_CAMPO",
  "message":"Falha na verificação do campo",
  "errors":[
    {
      "code":"CAMPO_NÃO_PODE_FICAR_VAZIO",
      "message":"Campo nome de usuário não pode ficar vazio"
    },{
      "code":"CAMPO_MUITO_LONGO",
      "message":"Campo senha está muito longo"
    },{
      "code":"ERRO_DESCONHECIDO",
      "message":"Erro Desconhecido"
    }
  ]
}
 */
```